src/cls/rbd/cls_rbd.cc

```diff
@@ -4797,6 +4803,246 @@ int image_get_group(cls_method_context_t hctx,
   return 0;
 }
 
+/**
+ * Retrieve next group snapshot id.
+ *
+ * Input:
+ * none
+ *
+ * Output:
+ * @param uint64_t
+ * @return 0 on success, negative error code on failure
+ */
+int group_snap_next_seq(cls_method_context_t hctx,
+			bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "group_snap_next_seq");
+  bufferlist seqbl;
+  int r = cls_cxx_map_get_val(hctx, GROUP_SNAP_SEQ, &seqbl);
+  if (r < 0 && r != -ENOENT) {
+    return r;
+  }
+
+  bufferlist::iterator iter = seqbl.begin();
+  uint64_t snap_seq = 0;
+  try {
+    ::decode(snap_seq, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  ::encode(snap_seq, *out);
+
+  return 0;
+}
+
+/**
+ * Save initial snapshot record.
+ *
+ * Input:
+ * @param GroupSnapshot
+ *
+ * Output:
+ * @return 0 on success, negative error code on failure
+ */
+int group_snap_save(cls_method_context_t hctx,
+		    bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "group_snap_save");
+  cls::rbd::GroupSnapshot gs;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(gs, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  bufferlist seqbl;
+  int r = cls_cxx_map_get_val(hctx, GROUP_SNAP_SEQ, &seqbl);
+  if (r < 0 && r != -ENOENT) {
+    return r;
+  }
+
+  bufferlist::iterator iter = seqbl.begin();
+  uint64_t snap_seq = 0;
+  try {
+    ::decode(snap_seq, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  // client lost a race with another snapshot creation.
+  // snap_seq must be monotonically increasing.
+  if (snap_seq > gs.id) {
+    return -ESTALE;
+  }
+
+  std::string key = gs.snap_key();
+
+  bufferlist obl;
+  ::encode(gs, obl);
+  r = cls_cxx_map_set_val(hctx, key, &obl);
+  if (r < 0) {
+    return r;
+  }
+
+  bufferlist sobl;
+  ++snap_seq;
+  ::encode(snap_seq, sobl);
+  r = cls_cxx_map_set_val(hctx, GROUP_SNAP_SEQ, &sobl);
+  return r;
+}
+
+/**
+ * Update snapshot record.
+ *
+ * Input:
+ * @param GroupSnapshot
+ *
+ * Output:
+ * @return 0 on success, negative error code on failure
+ */
+int group_snap_update(cls_method_context_t hctx,
+		      bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "group_snap_update");
+  cls::rbd::GroupSnapshot gs;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(gs, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  std::string key = gs.snap_key();
```

Comment:
Probably shouldn't allow all fields to be updated (e.g. the uuid shouldn't be updatable).